### PR TITLE
feat: performance results file validation

### PR DESCRIPTION
**Pull request by ShaswatPrabhat**
Validate performance results files input to compare package using zod

### CODE REVIEW

File: .changeset/angry-moles-bake.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@callstack/reassure-compare': minor
```
Comment (mdjastrzebski) [inline .changeset/angry-moles-bake.md:2]: This is also minor version bump for the main package and cli
```suggestion
'reassure': minor
'@callstack/reassure-cli': minor
'@callstack/reassure-compare': minor
```

File: .changeset/angry-moles-bake.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+---
+
+Validate performance results files input to compare package
```
Comment (mdjastrzebski) [inline .changeset/angry-moles-bake.md:5]:
```suggestion
Validate format of performance results files when loading for comparison
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+});
+
+const perfFileSchema = z
+  .array(performanceResultEntrySchema)
+  .refine((val) => !hasDuplicateValues(val.map((val) => val.name)), {
+    message: `Your performance result file contains records with duplicated names.
+    Please remove any non-unique names from your test suites and try again.`,
+  });
+
+export function validateAndParseMetadata(metadata: unknown) {
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:29]: We can use `parse` naming, validating parsed input is an expected thing.
```suggestion
export function parseMetadata(metadata: unknown) {
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { z } from 'zod';
+import { hasDuplicateValues } from './array';
+
+const metadataSchema = z.object({
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:4]: We would want to no longer have separate type definitions as we would need to manual sync these with zod. A better approach would be to use zod infering:

```suggestion
const performanceMetadataSchema = z.object({ ... });
type PerformanceMetadata = z.infer<typeof performanceMetadataSchema>;
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  metadata: z.object({
+    branch: z.string(),
+    commitHash: z.string(),
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:7]: As these are optional seems like we need to add `.optional()` chained method: https://github.com/colinhacks/zod#optionals

```suggestion
    branch: z.string().optional(),
    commitHash: z.string().optional(),
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  }),
+});
+
+const performanceResultEntrySchema = z.object({
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:11]: Lets use naming convention that uses type name and adds `Schema` to the end.
```suggestion
const PerformanceEntrySchema = z.object({
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  runs: z.number(),
+  meanDuration: z.number(),
+  stdevDuration: z.number(),
+  durations: z.array(z.number()),
+  meanCount: z.number(),
+  stdevCount: z.number(),
+  counts: z.array(z.number()),
+});
+
+const perfFileSchema = z
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:22]:
```suggestion
const PerformanceEntriesSchema = z
```

File: packages/reassure-compare/src/utils/validate.ts (hunk 6 above)
Comment (mdjastrzebski) [reply packages/reassure-compare/src/utils/validate.ts:11]: + add `z.infer<typeof PerformanceEntrySchema>;`

File: packages/reassure-compare/src/utils/validate.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  .refine((val) => !hasDuplicateValues(val.map((val) => val.name)), {
+    message: `Your performance result file contains records with duplicated names.
+    Please remove any non-unique names from your test suites and try again.`,
+  });
+
+export function validateAndParseMetadata(metadata: unknown) {
+  return metadataSchema.safeParse(metadata);
+}
+
+export function validateAndParsePerformanceEntries(perfEntry: unknown) {
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:33]:
```suggestion
export function parsePerformanceEntries(perfEntries: unknown) {
```

File: packages/reassure-compare/src/utils/validate.ts (hunk 4 above)
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:4]: This entry should be split into two:
* `PerformanceMetadataSchema` + infer
* `PerformanceHeaderSchema` + infer

Comment (ShaswatPrabhat) [reply packages/reassure-compare/src/utils/validate.ts:4]: We will export these types and use them in `types.ts` file right @mdjastrzebski ?

**Review by mdjastrzebski — CHANGES_REQUESTED** [review]
The code looks pretty good 👍🏻  I've added some comments about improving the code organisation a bit more.

I think we should also add some basic unit tests that would load a couple of hardcode performance files from disk and check that they match the expected output. The goal here is not not test Zod library, as it probably has it's own tests but rather to make sure that we are still able to run load typical performance results files. I would go with four input files: one with header, one without header, one with incorrect JSON, last one with duplicate entries.

File: packages/reassure-compare/src/utils/validate.ts (hunk 4 above)
Comment (ShaswatPrabhat) [reply packages/reassure-compare/src/utils/validate.ts:4]: We cannot replace `PerformanceMetadataSchema` directly as its present state does not have a top level `metadata` tag.

Comment (mdjastrzebski) [reply packages/reassure-compare/src/utils/validate.ts:4]: Let's add both `PerformanceMetadataScheme` (=== PerformanceMetadata type)
and `PerformanceHeaderScheme` (PerformanceHeader type)

**Comment by ShaswatPrabhat** [comment]
Have resolved all of the code feedback @mdjastrzebski .

Will get started on unit testing.

**Comment by ShaswatPrabhat** [comment]
@mdjastrzebski  will it be a good idea to `export` `loadFile` from `compare.ts` and write tests against it then?

Not very keen on writing tests for the full `compare` method.

Another approach can be to write tests for the exported methods from `validate.ts` file.

Please suggest.

**Comment by mdjastrzebski** [comment]
@ShaswatPrabhat exporting `loadFile` from file is ok, just do not add it to `index.ts` exports so that the whole package does not export it.

File: packages/reassure-compare/src/utils/validate.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
+import { z } from 'zod';
+import { hasDuplicateValues } from './array';
+
+const performanceMetadataSchema = z.object({
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:4]: I think it's good idea to move the schemes and `z.infers` to `types` as they are more type related.

File: packages/reassure-compare/src/utils/validate.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+const perfFileSchema = z
+  .array(performanceEntrySchema)
+  .refine((val) => !hasDuplicateValues(val.map((val) => val.name)), {
+    message: `Your performance result file contains records with duplicated names.
+    Please remove any non-unique names from your test suites and try again.`,
+  });
+
+export function parseMetadata(metadata: unknown) {
+  return performanceHeaderSchema.parse(metadata);
+}
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:54]:
```suggestion
export function parseHeader(header: unknown) {
  return performanceHeaderSchema.parse(header);
}
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+export function parsePerformanceEntries(perfEntry: unknown) {
+  return perfFileSchema.parse(perfEntry);
+}
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:58]:
```suggestion
export function parsePerformanceEntries(entries: unknown[]) {
  return perfFileSchema.parse(entries);
}
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  /** Standard deviation of measured render counts for each run */
+  stdevCount: z.number(),
+
+  /** Array of measured render counts for each run */
+  counts: z.array(z.number()),
+});
+
+export type PerformanceEntry = z.infer<typeof performanceEntrySchema>;
+
+const perfFileSchema = z
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:45]:
```suggestion
const PerformanceEntriesSchema = z
```

File: packages/reassure-compare/src/utils/validate.ts [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+export type PerformanceMetadata = z.infer<typeof performanceMetadataSchema>;
+
+const performanceHeaderSchema = z.object({
+  metadata: performanceMetadataSchema,
+});
+
+export type PerformanceHeader = z.infer<typeof performanceHeaderSchema>;
+
+const performanceEntrySchema = z.object({
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/utils/validate.ts:17]:
```suggestion
const PerformanceEntrySchema = z.object({
```

File: packages/reassure-compare/src/compare.ts [hunk 14]
```diff
@@ -58,8 +50,13 @@ export async function compare({
     logError(`Current results files "${currentFile}" does not exists. Check your setup.`);
     process.exit(1);
   }
-
-  const currentResults = await loadFile(currentFile);
+  let currentResults: PerformanceResults;
+  try {
+    currentResults = await loadFile(currentFile);
+  } catch (error) {
+    logError(`Error while parsing input file: ${currentFile}`, error);
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/compare.ts:57]: What is the error actually logged to console?

File: packages/reassure-compare/src/compare.ts [hunk 15]
```diff
@@ -68,7 +65,18 @@ export async function compare({
...
+  if (hasBaselineFile) {
+    try {
+      baselineResults = await loadFile(baselineFile);
+    } catch (error) {
+      logError(`Error while parsing input file: ${currentFile}`, error);
+      process.exit(1);
+    }
+  } else {
+    baselineResults = null;
+  }
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/compare.ts:79]: style:
```suggestion
  let baselineResults: PerformanceResults | null = null;
  if (hasBaselineFile) {
    try {
      baselineResults = await loadFile(baselineFile);
    } catch (error) {
      logError(`Error while parsing input file: ${currentFile}`, error);
      process.exit(1);
    }
  }
```

File: packages/reassure-compare/src/types.ts [hunk 16]
```diff
@@ -1,47 +1,11 @@
...
-
-/** Metadata information for performance results. */
-export interface PerformanceMetadata {
-  branch?: string;
-  commitHash?: string;
-}
-
-/** Entry in the performance results file. */
-export interface PerformanceEntry {
-  /** Name of the test scenario */
```
Comment (mdjastrzebski) [inline packages/reassure-compare/src/types.ts:20]: Can we save the code comments and moved them over respective Zod schema definitions for all the types and fields involved?

File: packages/reassure-compare/src/compare.ts (hunk 14 above)
Comment (ShaswatPrabhat) [reply packages/reassure-compare/src/compare.ts:64]: Will get the exact value in a test maybe? Otherwise will share from docs .

**Comment by ShaswatPrabhat** [comment]
Have added 2 basic snapshot tests based on `jest` and initial setup.

Please let me know if this is similar to what you have in mind @mdjastrzebski 

As the errors and success are both too large hence have snapshotted them

File: packages/reassure-compare/src/compare.ts (hunk 14 above)
Comment (ShaswatPrabhat) [reply packages/reassure-compare/src/compare.ts:64]: Right now it is something like:

[ZodError: [
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      0,
      "name"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      0,
      "runs"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      0,
      "meanDuration"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      0,
      "stdevDuration"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      0,
      "durations"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      0,
      "meanCount"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      0,
      "stdevCount"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      0,
      "counts"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      1,
      "name"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "runs"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "meanDuration"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "stdevDuration"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      1,
      "durations"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "meanCount"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      1,
      "stdevCount"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      1,
      "counts"
    ],
    "message": "Required"
  }
]]

**Review by mdjastrzebski — COMMENTED** [review]
@ShaswatPrabhat Looks good, I plan on merging it tomorrow. Thank you for contributing this feature.

**Review by mdjastrzebski — APPROVED** [review]
Looks great! Thanks for contributing this feature @ShaswatPrabhat 🚀

**Pull request merged**